### PR TITLE
fix: for token pair only quote against v2 quoter

### PR DESCRIPTION
**Pull request by jsy1218**
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bug fix

- **What is the current behavior?** (You can also link to an open issue here)
We will attempt to quote a FOT pair against v3 quoter or mixed quoter.

- **What is the new behavior (if this is a feature change)?**
FOT doesn't work in v3 pool. This means quoting against v3 quoter or mixed quoter (which goes through at least one v3 pool of the mixed route) doesn't work. If we see the tokenIn is sell FOT tax or tokenOut is buy FOT tax, then we only go through v2 quoter.

- **Other information**:
Testing wise, I'm relying on the existing e2e tests in both SOR and routing-api. On SOR level, it never goes through cached routes, so it will test against `swapRouteFromChain`. On routing-api level, we constantly run the code pipeline, so that the same FOT e2e test will likely go through the cached routes, i.e. `swapRouteFromCache`.

### CODE REVIEW

**Comment by jsy1218** [comment]
* **#677** <a href="https://app.graphite.dev/github/pr/Uniswap/smart-order-router/677?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @jsy1218 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/routers/alpha-router/alpha-router.ts [hunk 1]
```diff
@@ -2175,58 +2230,60 @@ export class AlphaRouter
...
-    // Maybe Quote mixed routes
-    // if MixedProtocol is specified or no protocol is specified and v2 is supported AND tradeType is ExactIn
-    // AND is Mainnet or Gorli
-    if (shouldQueryMixedProtocol && mixedProtocolAllowed) {
-      log.info({ protocols, tradeType }, 'Routing across MixedRoutes');
+    if (!fotInDirectSwap) {
+      // Maybe Quote mixed routes
+      // if MixedProtocol is specified or no protocol is specified and v2 is supported AND tradeType is ExactIn
+      // AND is Mainnet or Gorli
+      if (shouldQueryMixedProtocol && mixedProtocolAllowed) {
```
Comment (cgkol) [inline src/routers/alpha-router/alpha-router.ts:2237]: shall you add it in this if statement?+

File: src/routers/alpha-router/alpha-router.ts [hunk 2]
```diff
@@ -1733,6 +1748,22 @@ export class AlphaRouter
     swapConfig?: SwapOptions,
     providerConfig?: ProviderConfig
   ): Promise<BestSwapRoute | null> {
+    const tokenPairProperties =
```
Comment (cgkol) [inline src/routers/alpha-router/alpha-router.ts:1751]: nit: maybe lets make this a util function? seems like we call it a few times?

**Review by cgkol — COMMENTED** [review]
any tests we might want to add here to make sure this deos not break? i would suggest something like confirming that with fot buy ro sell we only get v2 candidate pools (good way to also proof v4 and future versions as well)

**Comment by jsy1218** [comment]
> any tests we might want to add here to make sure this deos not break? i would suggest something like confirming that with fot buy ro sell we only get v2 candidate pools (good way to also proof v4 and future versions as well)

potentially we can add tests in alpha-router.test.ts.

File: src/routers/alpha-router/alpha-router.ts (hunk 1 above)
Comment (jsy1218) [reply src/routers/alpha-router/alpha-router.ts:2237]: I think it's more readable to make a outer-scoped if statement for `fotInDirectSwap`.  Also there are some complex conditionals that makes it prone to make coding mistakes, e.g. 

```
if (
        v3ProtocolSpecified ||
        noProtocolsSpecified ||
        (shouldQueryMixedProtocol && mixedProtocolAllowed)
   ) {}
```

I can make it such that:

```
if (
        (v3ProtocolSpecified ||
        noProtocolsSpecified ||
        (shouldQueryMixedProtocol && mixedProtocolAllowed)) &&
       fotInDirectSwap
   ) {}

```

But it's more readable with:

```
if (!fotInDirectSwap) {
      if (
        v3ProtocolSpecified ||
        noProtocolsSpecified ||
        (shouldQueryMixedProtocol && mixedProtocolAllowed)
      ) {
       }
}
```

File: src/routers/alpha-router/alpha-router.ts (hunk 2 above)
Comment (jsy1218) [reply src/routers/alpha-router/alpha-router.ts:1751]: I tried it, and somehow some unit tests in `alpha-router.test.ts` started failing.

Comment (jsy1218) [reply src/routers/alpha-router/alpha-router.ts:1751]: I suspect it's mocking the token properties in those unit tests at the alpha-router class, which makes it kinda hard to do this kind of refactoring.

File: src/routers/alpha-router/alpha-router.ts (hunk 1 above)
Comment (cgkol) [reply src/routers/alpha-router/alpha-router.ts:2237]: fair enough

**Pull request merged**